### PR TITLE
remove checkpoint tracking from internal debugger

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Follow up to #9188 

Replaces the calls to the internal dev debugger in the model checkpoint. Tests are updated to cover what the "debugger" offered for tracking the calls inside the callback.
This is just one step towards removing the InternalDebugger completely in favor of proper unit testing.

### Does your PR introduce any breaking changes? If yes, please list them.

No, the "InternalDebugger" is internal as the name suggests.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: tests/checkpointing/test_checkpoint_callback_frequency.py [hunk 1]
```diff
@@ -17,34 +17,18 @@
...
-    # -----------------
-    val_train_model = BoringModel()
+def test_checkpoint_callback_disabled(tmpdir):
     # no callback
     trainer = Trainer(max_epochs=3, checkpoint_callback=False)
-    trainer.fit(val_train_model)
-    assert len(trainer.dev_debugger.checkpoint_callback_history) == 0
+    assert not any(isinstance(c, ModelCheckpoint) for c in trainer.callbacks)
+    trainer.fit(BoringModel())
+    assert not any(isinstance(c, ModelCheckpoint) for c in trainer.callbacks)
```
Comment (carmocca) [inline tests/checkpointing/test_checkpoint_callback_frequency.py:31]:
```suggestion
    assert not trainer.checkpoint_callbacks
    trainer.fit(BoringModel())
    assert not trainer.checkpoint_callbacks
```

File: tests/checkpointing/test_model_checkpoint.py [hunk 2]
```diff
@@ -769,11 +768,21 @@ def test_default_checkpoint_behavior(tmpdir):
...
+    assert save_mock.call_count == 3
+
+    save_dir = tmpdir / "lightning_logs" / "version_0" / "checkpoints"
+    save_mock.assert_has_calls(
+        [
+            call(trainer, save_dir / "epoch=0-step=4.ckpt"),
+            call(trainer, save_dir / "epoch=1-step=9.ckpt"),
+            call(trainer, save_dir / "epoch=2-step=14.ckpt"),
+        ]
+    )
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:785]: since you added this, remove the lines below?

```diff
-    # make sure the checkpoint we saved has the metric in the name
-    ckpts = os.listdir(os.path.join(tmpdir, "lightning_logs", "version_0", "checkpoints"))
-    assert len(ckpts) == 1
-    assert ckpts[0] == "epoch=2-step=14.ckpt"
```

**Review by carmocca — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by Borda — APPROVED** [review]

File: tests/checkpointing/test_model_checkpoint.py (hunk 2 above)
Comment (awaelchli) [reply tests/checkpointing/test_model_checkpoint.py:785]: I think we still want to assert that only the the latest one is saved and that the others are not present on the filesystem.

**Review by ananthsub — APPROVED** [review]

**Comment by carmocca** [comment]
There seem to be some failing tests for MacOS which have crept into master:
`tests/checkpointing/test_model_checkpoint.py::test_model_checkpoint_score_and_ckpt_val_check_interval`

**Comment by awaelchli** [comment]
I didn't notice.
But I haven't modified this test, have I?

**Comment by carmocca** [comment]
> But I haven't modified this test, have I?

Doesn't look like it, but if I'm not mistaken, this is the first PR to get the failures so I can only guess it comes from one of the changes...

**Comment by awaelchli** [comment]
They are not failing anymore on our latest runs on master: https://github.com/PyTorchLightning/pytorch-lightning/runs/3563724423

I don't know how to reproduce it or how my changes could have caused this.
